### PR TITLE
[12.0][FIX] contract: Modification mail + Auto-create follower in contract creation related to partner_id

### PR DESCRIPTION
**Pull request by victoralmau**
- [FIX] Modification mail to correct partner_ids (add `subtype_id` in function).
- [IMP] Improve tests
- [IMP] Auto-create follower related to `partner_id`
- [FIX] I have changed `_invoice_followers` function to other place to prevent add all followers from all contracts to all invoices created.

Please, @pedrobaeza and @chienandalu  review it.
@Tecnativa TT27605

### CODE REVIEW

**Review by pedrobaeza — COMMENTED** [review]
Can we remove then all the previous code checking if there are recipients to send?

File: contract/models/contract.py [hunk 1]
```diff
@@ -428,6 +428,7 @@ def _prepare_invoice(self, date_invoice, journal=None):
             'journal_id': journal.id,
             'origin': self.name,
             'user_id': self.user_id.id,
+            'old_contract_id': self.id
```
Comment (pedrobaeza) [inline contract/models/contract.py]: Please don't use this deprecated field. You need to inject the followers on the created invoices guessing them from the computed field, the same as the smart-button does.

File: contract/tests/test_contract.py [hunk 2]
```diff
@@ -248,15 +252,15 @@ def test_contract_invoice_followers(self):
         self.acct_line.recurring_rule_type = 'daily'
         self.contract.pricelist_id = False
         self.contract.message_subscribe(
-            partner_ids=self.contract.partner_id.ids,
-            subtype_ids=self.env.ref(
-                'contract.mail_message_subtype_invoice_created'
-            ).ids
+            partner_ids=[self.contract.partner_id.id],
+            subtype_ids=[
+                self.env.ref('contract.mail_message_subtype_invoice_created').id
```
Comment (pedrobaeza) [inline contract/tests/test_contract.py]: Why are you doing this change that doesn't bring anything?

File: contract/tests/test_contract.py [hunk 3]
```diff
@@ -248,15 +252,15 @@ def test_contract_invoice_followers(self):
...
+            ]
         )
         self.contract.recurring_create_invoice()
         invoice_daily = self.contract._get_related_invoices()
         self.assertTrue(invoice_daily)
-        self.assertEqual(len(invoice_daily.message_partner_ids.ids), 2)
+        self.assertEqual(len(invoice_daily.message_partner_ids.ids), 1)
```
Comment (pedrobaeza) [inline contract/tests/test_contract.py:263]: Why this change?

File: contract/tests/test_contract.py [hunk 4]
```diff
@@ -135,11 +135,11 @@ def _add_template_line(self, overrides=None):
...
 
     def test_add_modifications(self):
+        partner2 = self.partner.copy()
+        subtype = self.env.ref('contract.mail_message_subtype_contract_modification')
         self.contract.message_subscribe(
-            partner_ids=self.contract.partner_id.ids,
-            subtype_ids=self.env.ref(
-                'contract.mail_message_subtype_contract_modification'
-            ).ids
+            partner_ids=[partner2.id],
```
Comment (pedrobaeza) [inline contract/tests/test_contract.py]: Why are you preferring now this form instead of the more readable `partner_ids=partner2.ids`?

File: contract/tests/test_contract.py (hunk 3 above)
Comment (victoralmau) [reply contract/tests/test_contract.py:263]: IMO because before, more followers were added than they should.

**Review by pedrobaeza — APPROVED** [review]

File: contract/models/contract.py [hunk 5]
```diff
@@ -485,7 +485,10 @@ def _invoice_followers(self, invoices):
                 lambda x: invoice_create_subtype in x.subtype_ids
             ).mapped('partner_id')
             if partner_ids:
-                invoices.message_subscribe(partner_ids=partner_ids.ids)
+                item_invoices = item._get_related_invoices()
+                invoices.filtered(
+                    lambda x: item_invoices in x
```
Comment (pedrobaeza) [inline contract/models/contract.py]: Wait, this is not correct, as `get_related_invoices` can return several old invoices:
```suggestion
                    lambda x: bool(item_invoices & x)
```

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: contract/models/contract.py (hunk 5 above)
Comment (pedrobaeza) [reply contract/models/contract.py]: I have fixed it in a more elegant way:
```python
(invoices & item._get_related_invoices()).message_subscribe(...
```

**Review by pedrobaeza — APPROVED** [review]

**Comment by pedrobaeza** [comment]
Fast-tracking due to the severity of the bug:

/ocabot merge patch

**Comment by pedrobaeza** [comment]
Cherry-picked for v13 at ecacc4e8f249dd881fdda627972ae0f8a7fda87f

**Comment by StefanRijnhart** [comment]
This change seems to break my tests

```
2021-01-28 10:19:20,604 951171 ERROR 12contract odoo.addons.contract.tests.test_contract: FAIL: test_contract_invoice_followers (odoo.addons.contract.tests.test_contract.TestContract) 
2021-01-28 10:19:20,605 951171 ERROR 12contract odoo.addons.contract.tests.test_contract: Traceback (most recent call last): 
2021-01-28 10:19:20,605 951171 ERROR 12contract odoo.addons.contract.tests.test_contract: `   File "contract/contract/tests/test_contract.py", line 264, in test_contract_invoice_followers 
2021-01-28 10:19:20,605 951171 ERROR 12contract odoo.addons.contract.tests.test_contract: `     self.assertTrue(invoice_daily) 
2021-01-28 10:19:20,605 951171 ERROR 12contract odoo.addons.contract.tests.test_contract: ` AssertionError: 2 != 1 
```

Subscribed users are 

```
(Pdb) invoice_daily.message_partner_ids.mapped('name')
['OdooBot', 'partner test contract']
```

Is this what you would expect?

OdooBot is the creator of the invoice and as such is created here: https://github.com/odoo/odoo/blob/12.0/addons/mail/models/mail_thread.py#L269.

This set of message partners is the same in the test before this PR was merged. I see this PR was merged when Travis was disfunctional and tests were not always run properly, but it looks like there was at least one successful travis run with this code (https://travis-ci.com/github/OCA/contract/jobs/472018591#L2078).

I'm just going to add a PR with this test change reverted and see what happens.
